src/libcore/char.rs

```diff
@@ -448,19 +459,15 @@ impl Iterator for EscapeUnicode {
                 Some('u')
             }
             EscapeUnicodeState::LeftBrace => {
-                let mut n = 0;
-                while (self.c as u32) >> (4 * (n + 1)) != 0 {
-                    n += 1;
-                }
-                self.state = EscapeUnicodeState::Value(n);
+                self.state = EscapeUnicodeState::Value;
                 Some('{')
             }
-            EscapeUnicodeState::Value(offset) => {
-                let c = from_digit(((self.c as u32) >> (offset * 4)) & 0xf, 16).unwrap();
-                if offset == 0 {
+            EscapeUnicodeState::Value => {
+                let c = from_digit(((self.c as u32) >> (self.offset * 4)) & 0xf, 16).unwrap();
```

Comment:
I do not get why won’t you mutate the `self.c` instead of book-keeping the `offset` thing.

i.e. you could do something like this

```rust
// in escape_unicode
while self.c & 0xf000_0000 == 0 { self.c = self.c << 2 } // or an equivalent version with `leading_zeroes`
// here
let nibble = self.c & 0xf000_0000;
self.c = self.c << 2;
if self.c == 0 { self.state = RightBrace }
Some(from_digit(nibble))
```

Such approach would save you complexity of implementation, space in the iterator structure and perhaps make the code faster?